src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java

```diff
@@ -0,0 +1,147 @@
+     */
+    public HttpUrl getBaseUrl() {
+        return bitbucketBaseUrl;
+    }
+
+    /**
+     * Provide the base Rest path
+     *
+     * @return HttpUrl of the core rest path;
+     */
+    public HttpUrl getCoreRestPath() {
+        return coreRestPathUrl;
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * If the requested resource is paged, or the return type is generified use this method,
+     * otherwise the {@link #makeGetRequest(HttpUrl, Class)} is most likely a better choice.
+     *
+     * @param url        url to connect to
+     * @param returnType type reference used when getting generified objects (such as pages)
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, Class)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, TypeReference<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * <em>Note!</em> this method <em>cannot</em> be used to retrieve entities that makes use of
+     * generics (such as {@link BitbucketPage}) for that use {@link #makeGetRequest(HttpUrl,
+     * TypeReference)} instead.
+     *
+     * @param url        url to connect to
+     * @param returnType class of the desired return type. Do note that if the type is generified
+     *                   this method will not work
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, TypeReference)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, Class<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Makes a POST request to the given URL with given request payload.
+     *
+     * @param url,            the URL to make the request to
+     * @param requestPayload, JSON payload which will be marshalled to send it with POST.
+     * @param returnType,     Class of expected return type
+     * @param <T>,            Type of Request payload.
+     * @param <R>,            Return type
+     * @return
+     */
+    public <T, R> BitbucketResponse<R> makePostRequest(HttpUrl url, T requestPayload, Class<R> returnType) {
+        ObjectReader<R> reader = in -> objectMapper.readValue(in, returnType);
+        return httpRequestExecutor.executePost(url, credentials, marshall(requestPayload), response ->
+                new BitbucketResponse<>(response.headers().toMultimap(), unmarshall(reader, response.body())));
+    }
+
+    private void ensureNonEmptyBody(Response response) {
+        if (response.body() == null) {
+            log.debug("Bitbucket - No content in response");
+            throw new NoContentException(
```

Comment:
More a question than a reviewer comment, but is a 204 necessarily 'exceptional behaviour' for a `GET` call? If the client anticipates no content, can they handle that case on their end?

Reply:
204 for a `GET` would be unusual, while technically allowed. For this I would re-iterate my point that we're not building a generic HTTP nor a generic Bitbucket Server client.  We're building one for a specific purpose, so I think this is fine, while technically this _could_ cause a problem, but in practise it will not as Bitbucket Server only returns a 204 under very specific circumstances which we can handle if we need to.